File: 3104-find-longest-self-contained-substring/3104-find-longest-self-contained-substring.cpp
class Solution {
public:
    int maxSubstringLength(string s) {
        int n = s.length();
        map<int, int> first; map<int, int> last;
        for(int i = 0; i < n; ++i) {
            if(first.find(s[i]-'a') == first.end()) first[s[i]-'a'] = i;
            last[s[i] - 'a'] = i;
        }
        int ans = -1;
        for(auto& [c, i] : first) {
            int mx = last[c];
            for(int j = i; j < n; ++j) {
                int a = first[s[j]-'a']; int b = last[s[j]-'a'];
                if(a < i) break;
                mx = max(mx, b);
                if(mx == j && j - i + 1 < n) ans = max(ans, j - i + 1);
            }
        }
        return ans;
    }
};

File: 3104-find-longest-self-contained-substring/3104-find-longest-self-contained-substring.py
class Solution(object):
    def maxSubstringLength(self, s):
        """
        :type s: str
        :rtype: int
        """
        ans, n = -1, len(s)
        first, last = {}, {}
        for i, c in enumerate(s):
            if c not in first: first[c] = i
            last[c] = i
        
        for c, i in first.items():
            mx = last[c]
            for j in range(i, n):
                a, b = first[s[j]], last[s[j]]
                if a < i: break
                mx = max(mx, b)
                if mx == j and j - i + 1 < n:
                    ans = max(ans, j - i + 1)
        return ans